go/cmd/dolt/commands/sqlserver/server.go

```diff
@@ -63,335 +64,470 @@ func Serve(
+					Handler: mux,
+				}
 
-		go func() {
-			_ = metSrv.ListenAndServe()
-		}()
+				_ = metSrv.ListenAndServe()
+			}
+		},
+		Stop: func() error {
+			if metSrv != nil {
+				metSrv.Close()
+			}
+			return nil
+		},
 	}
+	controller.Register(RunMetricsServer)
 
 	var remoteSrv *remotesrv.Server
-	if serverConfig.RemotesapiPort() != nil {
-		port := *serverConfig.RemotesapiPort()
-		listenaddr := fmt.Sprintf(":%d", port)
-		args, err := sqle.RemoteSrvServerArgs(sqlEngine.NewDefaultContext, remotesrv.ServerArgs{
-			Logger:         logrus.NewEntry(lgr),
-			ReadOnly:       true,
-			HttpListenAddr: listenaddr,
-			GrpcListenAddr: listenaddr,
-		})
-		if err != nil {
-			lgr.Errorf("error creating SQL engine context for remotesapi server: %v", err)
-			startError = err
-			return
-		}
+	var remoteSrvListeners remotesrv.Listeners
+	RunRemoteSrv := &svcs.Service{
+		Init: func(ctx context.Context) error {
+			if serverConfig.RemotesapiPort() == nil {
+				return nil
+			}
+
+			port := *serverConfig.RemotesapiPort()
+			listenaddr := fmt.Sprintf(":%d", port)
+			args, err := sqle.RemoteSrvServerArgs(sqlEngine.NewDefaultContext, remotesrv.ServerArgs{
+				Logger:         logrus.NewEntry(lgr),
+				ReadOnly:       true,
+				HttpListenAddr: listenaddr,
+				GrpcListenAddr: listenaddr,
+			})
+			if err != nil {
+				lgr.Errorf("error creating SQL engine context for remotesapi server: %v", err)
+				return err
+			}
 
-		ctxFactory := func() (*sql.Context, error) { return sqlEngine.NewDefaultContext(ctx) }
-		authenticator := newAuthenticator(ctxFactory, sqlEngine.GetUnderlyingEngine().Analyzer.Catalog.MySQLDb)
-		args = sqle.WithUserPasswordAuth(args, authenticator)
+			ctxFactory := func() (*sql.Context, error) { return sqlEngine.NewDefaultContext(ctx) }
+			authenticator := newAuthenticator(ctxFactory, sqlEngine.GetUnderlyingEngine().Analyzer.Catalog.MySQLDb)
+			args = sqle.WithUserPasswordAuth(args, authenticator)
+			args.TLSConfig = serverConf.TLSConfig
 
-		args.TLSConfig = serverConf.TLSConfig
-		remoteSrv, err = remotesrv.NewServer(args)
-		if err != nil {
-			lgr.Errorf("error creating remotesapi server on port %d: %v", port, err)
-			startError = err
-			return
-		}
-		listeners, err := remoteSrv.Listeners()
-		if err != nil {
-			lgr.Errorf("error starting remotesapi server listeners on port %d: %v", port, err)
-			startError = err
-			return
-		} else {
-			go remoteSrv.Serve(listeners)
-		}
+			remoteSrv, err = remotesrv.NewServer(args)
+			if err != nil {
+				lgr.Errorf("error creating remotesapi server on port %d: %v", port, err)
+				return err
+			}
+			remoteSrvListeners, err = remoteSrv.Listeners()
+			if err != nil {
+				lgr.Errorf("error starting remotesapi server listeners on port %d: %v", port, err)
+				return err
+			}
+			return nil
+		},
+		Run: func(ctx context.Context) {
+			if remoteSrv == nil {
+				return
+			}
+			remoteSrv.Serve(remoteSrvListeners)
+		},
+		Stop: func() error {
```

Comment:
TODO: There's a state where Init() and Stop() get called but Run() never gets called. In that case, the grpc.Server instance never took ownership of the listeners and we should close them here.

AFAICT, to correctly handle this would require some thread-safe coordination between `Run` and `Stop`, so that `Stop` took responsibility for the listeners only if Serve had never been called, and otherwise the listeners were handled as they are now, by `GracefulStop`.

This state also exists in `RunClusterRemoteSrv` just below.